extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java

```diff
@@ -80,7 +83,17 @@ public OpenCensusProtobufReader(
   @Override
   public CloseableIterator<InputRow> read()
   {
-    return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+    Supplier<Iterator<InputRow>> supplier = Suppliers.memoize(() -> readAsList().iterator());
+    return CloseableIterators.withEmptyBaggage(new Iterator<InputRow>() {
```

Comment:
I don't think this code is obvious at all, we should add some comments to explain why this is necessary

Reply:
also, rather than repeating this code twice it might be worth having a utility method do to this in both readers